tools/myst/enc/enc.c

```diff
@@ -143,37 +125,60 @@ static uint64_t _forward_exception_as_signal_to_kernel(
         // Note that the si_addr in the simulated #PF always has the
         // lower 12 bits cleared.
         siginfo.si_addr = (void*)oe_exception_record->faulting_address;
-        (*_kargs.myst_handle_host_signal)(&siginfo, &_mcontext);
-        _mcontext_to_oe_context(&_mcontext, oe_context);
-        return OE_EXCEPTION_CONTINUE_EXECUTION;
+        supported_code = true;
     }
-    if (oe_exception_code == OE_EXCEPTION_ACCESS_VIOLATION)
+    else if (oe_exception_code == OE_EXCEPTION_ACCESS_VIOLATION)
     {
         // #GP can only be delivered on icelake.
         siginfo.si_code = SEGV_ACCERR;
         siginfo.si_signo = SIGSEGV;
         // `si_addr` is always null for #GP.
         siginfo.si_addr = NULL;
-        (*_kargs.myst_handle_host_signal)(&siginfo, &_mcontext);
-        _mcontext_to_oe_context(&_mcontext, oe_context);
-        return OE_EXCEPTION_CONTINUE_EXECUTION;
+        supported_code = true;
     }
-    if (oe_exception_code == OE_EXCEPTION_X87_FLOAT_POINT)
+    else if (oe_exception_code == OE_EXCEPTION_X87_FLOAT_POINT)
     {
         // ATTN: Consider implementing accurate si-code for
         // OE_EXCEPTION_X87_FLOAT_POINT
         siginfo.si_code = FPE_FLTINV;
         siginfo.si_signo = SIGFPE;
-        (*_kargs.myst_handle_host_signal)(&siginfo, &_mcontext);
-        _mcontext_to_oe_context(&_mcontext, oe_context);
-        return OE_EXCEPTION_CONTINUE_EXECUTION;
+        supported_code = true;
     }
-    if (oe_exception_code == OE_EXCEPTION_DIVIDE_BY_ZERO)
+    else if (oe_exception_code == OE_EXCEPTION_DIVIDE_BY_ZERO)
     {
         siginfo.si_code = FPE_INTDIV;
         siginfo.si_signo = SIGFPE;
-        (*_kargs.myst_handle_host_signal)(&siginfo, &_mcontext);
-        _mcontext_to_oe_context(&_mcontext, oe_context);
+        supported_code = true;
+    }
+
+    if (supported_code)
+    {
+        /* The OE layer hardware exception handling scheme requires
+         * any registered handler call to return such that the OE
+         * can clean up internal states after the handler finishes.
+         * However, invoking a signal handler of the programming
+         * language or the application does not always meet the
+         * requirement as the handler might not return. Therefore,
```

Comment:
If the handler does not return, then the control won't reach here. It is not clear, then, why this code is needed.

Reply:
With this PR, we do not invoke use handler code before this point (instead after this function returns)